Several fixes

Giperion opened the pull request:
Fixed dereferencing nullptr, when sending CMSG_LOOT guid with HIGHGUID_GAMEOBJECT, but invalid in general.
Prevent cheating with LANG_UNIVERSAL and message size.
Don't allow cast "Honorless target" to self
Warden fixes.

Giperion committed:
* Fixed dereferencing nullptr, when sending CMSG_LOOT guid with HIGHGUID_GAMEOBJECT, but invalid in general.

Giperion committed:
* Don't allow sending LANG_UNIVERSAL by player
* Don't allow messages with more that 255 symbols (server side check)

Giperion committed:
* Prevent cheaters cast honorless target on self

Giperion committed:
* Now warden reacts on client delay when initialization in progress or strange behaviour with initialization logic

Giperion committed:
* Fix for previous commit: Made a error while staging changes in git

billy1arm committed:
Update Player.cpp

billy1arm committed:
Update Warden.cpp

Rochet2 reviewed (commented):
Could all tabs be changed to 4 spaces like all other code has? Thanks :)

billy1arm committed:
Update Warden.cpp

billy1arm committed:
Update Warden.h

billy1arm committed:
Update WardenMac.cpp

billy1arm committed:
Update Warden.h

billy1arm committed:
Update WardenMac.cpp

billy1arm committed:
Update WardenWin.cpp

billy1arm committed:
Update ChatHandler.cpp

billy1arm committed:
Update Spell.cpp

Rochet2 commented on src/game/WorldHandlers/ChatHandler.cpp:
```diff
@@ -93,7 +100,10 @@ void WorldSession::HandleMessagechatOpcode(WorldPacket& recv_data)
+
+    //prevent cheating, by sending LANG_UNIVERSAL
+    if ((langDesc->lang_id == LANG_UNIVERSAL && !sWorld.getConfig(CONFIG_BOOL_ALLOW_TWO_SIDE_INTERACTION_CHAT) && GetSecurity() == SEC_PLAYER) ||
+         langDesc->skill_id != 0 && !_player->HasSkill(langDesc->skill_id))
```
I think here there should be brackets `()` to separate the || and && conditions in the way that they are intended to clarify what is meant. Currently it looks like `if ((A && !B && C) || D && !E)`

Giperion replied:
For separation i just use new line.
I think brackets here can only serve as noise.

Rochet2 replied:
Here is the code as it is at the moment `(A || B && C)` did you mean `((A || B) && C)` or did you mean `(A || (B && C))`?

Here is an example of a compiler warning generated by such code: https://travis-ci.org/Rochet2/smallfolk_cpp/jobs/202645497#L314-L317

billy1arm replied:
@Giperionn - any ideas on what Rochet2 mentioned above

Giperion replied:
I don't understand him. But i explain why i write this code.

if ((langDesc->lang_id == LANG_UNIVERSAL && !sWorld.getConfig(CONFIG_BOOL_ALLOW_TWO_SIDE_INTERACTION_CHAT) && GetSecurity() == SEC_PLAYER) ||
langDesc->skill_id != 0 && !_player->HasSkill(langDesc->skill_id))

if ((CondTier1_1 && !CondTier1_2 && CondTier1_3) ||
CondTier2_1 && !CondTier2_2)

If all conditions of Tier 1 is a true, we don't need check Tier 2 and can switch execution routine. This work, because all Tier1 conditions wrapped by brackets AND separated by newline from Tier 2 conditions. Brackets serve as micro function, returning true of false, depends on Tier 1 conditions.
If one of the conditions if Tier 1 is a false, then we check Tier 2 conditions, which finally decides can we switch execution routine or not.

Rochet2 commented on src/game/WorldHandlers/ChatHandler.cpp:
```diff
@@ -103,7 +113,7 @@ void WorldSession::HandleMessagechatOpcode(WorldPacket& recv_data)
         // Disabled addon channel?
         if (!sWorld.getConfig(CONFIG_BOOL_ADDON_CHANNEL))
-            { return; }
+        { return; }
```
indentation off

Rochet2 commented on src/game/Warden/Warden.h:
```diff
@@ -95,6 +95,43 @@ struct WardenHashRequest
+            case WardenState::STATE_RESTING:
+                return "STATE_RESTING";
+            default:
+                return "UNDEFINED STATE";
```
I think a better way is to leave the default case out and have a return in the end. This is because compilers can detect and complain if you have not defined all cases if you dont have a default case and we could maybe see if there is a missing warden tostring definition. But thats my opinion.

Giperion replied:
I write similiar code in my UE4 project (which work on Win, Android, iOS) and never got errors/strange behaviours from that.
At least MSVC and Clang work perfectly with that.

Rochet2 replied:
I didnt mean the code was wrong. Just that compilers can give warnings if cases are missing. This can lead to easier detection of errors.

Here is an example:
No default case: https://github.com/Rochet2/smallfolk_cpp/commit/31dd3a133b6b14054640c9912cc7e16db7f4cefa
Added new enum value without changing the switch: https://github.com/Rochet2/smallfolk_cpp/commit/802b86c8c3499db098c5a6a2fc697913812a64c3
And travis compile warnings (Im using the -Wall flag): https://travis-ci.org/Rochet2/smallfolk_cpp/jobs/202598262#L305-L308

Giperion replied:
All that i know - this code is completely valid. I write this and clang and MSVC understand this.

Giperion commented:
To be honest, i really don't know about spaces and tabs, until i got huge change list in SmartGit about them.
That's because Visual Studio automatically swap spaces to tabs.
I spend 20 min. in smartgit to undo that visual studio shit, but i can't fix my lines so easily, because if i change one line - visual studio do their duty work again :) .

Giperion commented:
By the way, thank billy1arm for fixing that tab/spaces issue.

billy1arm commented:
@Giperionn - which version of visual studio are you using, is I can guide you on changing the tabs setting in visual studio

Giperion commented:
For editing ManGOS and WoW related things i use 2015 version.

billy1arm commented:
@Giperionn - Here's how to configure VS correctly....
Select the Options Menu
Expand "Text Editor"
Expand "All Languages"
Expand "Tabs"
Select "Smart"
Select "Insert Spaces"

Click OK and you should be sorted

Rochet2 committed:
Remove default case from switch

This allows compilers to detect if a case for the WardenState enum is missing when using -Wall flag

Rochet2 committed:
Make the intention of the IF case more clear

Make the intention of the IF case more clear and suppress a compiler warning when using -Wall or /W4 flags
